Disabled gpu

brightskye opened the pull request:
I tried my best to do some unit test on a docker machine (without nvidia card of course) but will really appreciate if someone can help me test it. Thanks.

brightskye committed:
Merge commit '3b8b20f4290280e54d5c2b9f472e0c1b300ce12d' into blacklist-gpu

brightskye committed:
remove test file

brightskye committed:
Merge branch '19-2.1' into blacklist-gpu

brightskye committed:
renamed blacklist_gpu to disabled_gpu
added support on 6tempcontrol
added echo for disabled_gpu on watchdog and tempcontrol

brightskye committed:
merge from 19-2.1

brightskye committed:
merge from 19-2.1

brightskye committed:
removed uneeded comments

papampi commented:
Will test soon.

papampi commented:
First test with AUTO_SWITCH="WTM_SWITCHING"
Guake WTM_SWITCHING tab keeps repeating 
`Mon Jul 23 01:41:23 +0430 2018 - [INFO] - GPU1 disabled
`

Disabled switching and set it to "NO"now temp control only shows it

```
      TEMP_CONTROL v0019-2.1.0006 by leenoox      

Detected: 3 GPU's

GPU 0:  POWER LIMIT: 100,  TARGET TEMP: 70
GPU 1:  POWER LIMIT: 75,  TARGET TEMP: 70
GPU 2:  POWER LIMIT: 75,  TARGET TEMP: 70

GLOBAL  FAN_ADJUST (%):  2
GLOBAL  POWER_ADJUST (W):  2
GLOBAL  ALLOWED_TEMP_DIFF (C):  2
GLOBAL  RESTORE_POWER_LIMIT (%):  80
GLOBAL  MINIMAL_FAN_SPEED (%):  50

GPU 0, Target temp: 70, Current: 55, Diff: 15, Fan: 50, Power: 47.14
Mon Jul 23 01:43:04 +0430 2018 - [INFO] - GPU1 disabled
Mon Jul 23 01:43:04 +0430 2018 - [INFO] - GPU1 disabled
Mon Jul 23 01:43:04 +0430 2018 - [INFO] - GPU1 disabled
Mon Jul 23 01:43:04 +0430 2018 - [INFO] - GPU1 disabled
Mon Jul 23 01:43:04 +0430 2018 - [INFO] - GPU1 disabled
Mon Jul 23 01:43:04 +0430 2018 - [INFO] - GPU1 disabled
Mon Jul 23 01:43:04 +0430 2018 - [INFO] - GPU1 disabled
Mon Jul 23 01:43:04 +0430 2018 - [INFO] - GPU1 disabled
Mon Jul 23 01:43:04 +0430 2018 - [INFO] - GPU1 disabled
Mon Jul 23 01:43:04 +0430 2018 - [INFO] - GPU1 disabled
Mon Jul 23 01:43:04 +0430 2018 - [INFO] - GPU1 disabled
Mon Jul 23 01:43:04 +0430 2018 - [INFO] - GPU1 disabled
Mon Jul 23 01:43:04 +0430 2018 - [INFO] - GPU1 disabled
Mon Jul 23 01:43:04 +0430 2018 - [INFO] - GPU1 disabled
```

papampi commented:
And an error in watchdog:

```Mon Jul 23 01:42:25 +0430 2018 - [INFO] - Watchdog is starting

/home/m1/NVOC/disabled_gpu/5watchdog: line 138: ACTIVE_GPU_COUNT: command not found

Mon Jul 23 01:42:27 +0430 2018 - [INFO] - Installed GPUs: 3, Connected GPUs: 3

Connection to google.com 443 port [tcp/https] succeeded!
Giving miner enough to stabilize
Mon Jul 23 01:42:42 +0430 2018 - [INFO] - Miner is running: going on watch
Load Average: 1.00, 0.75, 0.61
Mon Jul 23 01:42:52 +0430 2018 - [INFO] - GPU0 Utilization: 100%
Mon Jul 23 01:42:53 +0430 2018 - [INFO] - GPU1 disabled
Mon Jul 23 01:42:53 +0430 2018 - [INFO] - GPU2 Utilization: 100%
Mon Jul 23 01:42:53 +0430 2018 - [INFO] - No mining issues detected, Check again in 15
 
Load Average: 1.12, 0.80, 0.63
Mon Jul 23 01:43:08 +0430 2018 - [INFO] - GPU0 Utilization: 0%
Mon Jul 23 01:43:08 +0430 2018 - [WARNING] - GPU0 under 70% threshold - UTILIZATION: 0%
Mon Jul 23 01:43:09 +0430 2018 - [INFO] - GPU1 disabled
Mon Jul 23 01:43:09 +0430 2018 - [INFO] - GPU2 Utilization: 0%
Mon Jul 23 01:43:09 +0430 2018 - [WARNING] - GPU2 under 70% threshold - UTILIZATION: 0%
/home/m1/NVOC/disabled_gpu/5watchdog: line 483: 100 * NUM_GPU_BLW_THRSHLD / ACTIVE_GPU_COUNT : division by 0 (error token is "ACTIVE_GPU_COUNT ")
```

papampi commented:
And if you disable gpu 0 the log wont show GPU0 but only that GPU is disabled:

`Mon Jul 23 02:09:25 +0430 2018 - [INFO] - GPU disabled`

My test rig only has 3 cards, so cant test +10 cards

papampi commented on 5watchdog:
```diff
@@ -134,11 +135,12 @@ fi
 REBOOTRESET=0
 GPU_COUNT=$(nvidia-smi -i 0 --query-gpu=count --format=csv,noheader,nounits)
-COUNT=$((6 * GPU_COUNT))
+ACTIVE_GPU_COUNT = $(($GPU_COUNT - ${#DISABLED_GPU_ARRAY[@]}))
```
I think line 138 should be:
ACTIVE_GPU_COUNT=$(($GPU_COUNT - ${#DISABLED_GPU_ARRAY[@]}))

brightskye replied:
Just fixed the ACTIVE_GPU_COUNT, 
wonder if this is the cause of the GPU 0 not showing up issue.

brightskye committed:
fix 6temp infinite loop due to not ++GPU ++count in the while loops
fix ACTIVE_GPU_COUNT assignment in watchdog

brightskye commented:
I think I have fixed the 6temp loop issues and the 
ACTIVE_GPU_COUNT=$(($GPU_COUNT - ${#DISABLED_GPU_ARRAY[@]})) , it's sitting in my repo, but i have no idea how to merge it here or how to proceed.

Sorry that I'm a total newbie when it comes to github -.-!
Any help is appreciated.

Thanks,

LuKePicci commented:
You just need to push new commits on this same branch, will appear here automatically.

brightskye commented:
I have already push it to my repo same blacklist-gpu branch.
However I haven't merge it with the latest change on 19-2.1, do i need to merge it first?

brightskye committed:
merge from 19-2.1

LuKePicci commented:
I can't see any new commit in your remote blacklist-gpu branch on github, have you just commited or pushed also?

Make those commits appear here, then you will take care of merging.

brightskye committed:
Merge branch 'blacklist-gpu' into disabled-gpu

brightskye commented:
nvm, I found what's the problem...
I got myself confused with an older branch that i forgot to delete.... -.-!

papampi commented:
watchdog is fixed, but tempcontrol has some problems.
If GPU0 disabled, first it opens nvidia-settings and stays on:
```
      TEMP_CONTROL v0019-2.1.0006 by leenoox

Detected: 3 GPU's

Mon Jul 23 19:24:57 +0430 2018 - [INFO] - GPU disabled
Mon Jul 23 19:24:57 +0430 2018 - [INFO] - GPU disabled
Mon Jul 23 19:24:57 +0430 2018 - [INFO] - GPU disabled
```
When you close nvidia-settings it starts working but with problems 
 ```

GLOBAL  FAN_ADJUST (%):  2
GLOBAL  POWER_ADJUST (W):  2
GLOBAL  ALLOWED_TEMP_DIFF (C):  2
GLOBAL  RESTORE_POWER_LIMIT (%):  80
GLOBAL  MINIMAL_FAN_SPEED (%):  50

Mon Jul 23 19:35:52 +0430 2018 - [INFO] - GPU0 disabled
GPU 1, Target temp: , Current: 59, Diff: -59, Fan: 58, Power: 74.08
GPU 1, Mon Jul 23 19:35:52 +0430 2018 - Adjusting fan from: 58 to: 62 Temp: 59

GPU 2, Target temp: , Current: 66, Diff: -66, Fan: 54, Power: 71.78
GPU 2, Mon Jul 23 19:35:53 +0430 2018 - Adjusting fan from: 54 to: 58 Temp: 66

Mon Jul 23 19:35:54 +0430 2018 - All good, will check again in 18.5 seconds



Mon Jul 23 19:36:12 +0430 2018 - [INFO] - GPU0 disabled
GPU 1, Target temp: , Current: 59, Diff: -59, Fan: 62, Power: 73.23
GPU 1, Mon Jul 23 19:36:12 +0430 2018 - Adjusting fan from: 62 to: 66 Temp: 59

GPU 2, Target temp: , Current: 66, Diff: -66, Fan: 58, Power: 74.64
GPU 2, Mon Jul 23 19:36:13 +0430 2018 - Adjusting fan from: 58 to: 62 Temp: 66

Mon Jul 23 19:36:14 +0430 2018 - All good, will check again in 18.5 seconds



Mon Jul 23 19:36:32 +0430 2018 - [INFO] - GPU0 disabled
GPU 1, Target temp: , Current: 59, Diff: -59, Fan: 66, Power: 74.58
GPU 1, Mon Jul 23 19:36:32 +0430 2018 - Adjusting fan from: 66 to: 70 Temp: 59

GPU 2, Target temp: , Current: 66, Diff: -66, Fan: 62, Power: 72.46
GPU 2, Mon Jul 23 19:36:33 +0430 2018 - Adjusting fan from: 62 to: 66 Temp: 66

Mon Jul 23 19:36:34 +0430 2018 - All good, will check again in 18.5 seconds

```

It get the temp diff wrong and start raising fan speed ...

papampi commented on 6tempcontrol:
```diff
@@ -214,6 +221,12 @@ do
   GPU=0
   while (( GPU < GPUS ))
```
What if we change its approach and use same as watchdog?
For example change the while loop to:
 for (( GPU=0; GPU < GPU_COUNT; GPU++))

LuKePicci replied:
Would be the very same thing.

papampi replied:
But it's more eye catching ... ;)
I was thinking to unify variables and arguments and ... So both use same system. for example watchdog uses GPU_COUNT while temp control use GPUS.

LuKePicci replied:
Yeah, uniforming names is always a good thing.

leenoox commented:
You will need to add new check inside the loop, something like:
If GPU$GPU == disabled then GPU++

then it will skip that one and go to the next.
I might be wrong about the variable name but you'll get the idea.

Sorry for not helping with this release, I honestly have no time at all to
pitch in... i am checking mail for all the changes you make and I can say
that you guys are doing great job. Lots of improvements :)


On Mon, Jul 23, 2018, 12:01 PM Luca Piccirillo, <notifications@github.com>
wrote:

> *@LuKePicci* commented on this pull request.
> ------------------------------
>
> In 6tempcontrol
> <https://github.com/papampi/nvOC_by_fullzero_Community_Release/pull/155#discussion_r204458187>
> :
>
> > @@ -214,6 +221,12 @@ do
>    GPU=0
>    while (( GPU < GPUS ))
>
> Yeah, uniforming names is always a good thing.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/papampi/nvOC_by_fullzero_Community_Release/pull/155#discussion_r204458187>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Ags0B3uC66HeOq7Fe99-94tOXnf9oU0Iks5uJfNTgaJpZM4VaH3G>
> .
>

papampi commented:
@leenoox Glad to see you bump in once in a awhile

brightskye committed:
fix incorrect loop index used for checking disabled gpu

brightskye commented:
Should be fixed now
I missed the $count in line 125, and was using $GPU which is an unset variable,
then found out array with unset variable always return the first value, thus causing it to skip all individual power , fan settings
which then results in $__CTL_ENABLE to be unset, and causing nvidia-settings to stay open waiting for command
Without the correct fan and powerlimit varible assigned correctly, everything becomes a mess -.-!

brightskye committed:
format 6temp disabled message

papampi commented on 6tempcontrol:
```diff
@@ -121,6 +122,12 @@ count=0
 do
+  if [ ${DISABLED_GPU_ARRAY[$count]} ]; then
+    # skip if disabled
+    echo "${B}GPU $count$:{N} ${R}Disabled"
```
Line 127 should be changed to:
echo -e "${B}GPU $count:${N}  ${R}${B}Disabled${N}"

papampi commented on 6tempcontrol:
```diff
@@ -214,6 +221,12 @@ do
   do
+    if [ ${DISABLED_GPU_ARRAY[$GPU]} ]; then
+      # skip if disabled
+      echo "${B}GPU $GPU -${N} ${R}Disabled"
```
Line 226 should be changed to:
echo -e "${B}GPU $GPU,${N} ${R}${B}Disabled${N}"

papampi commented:
All the rest looks good with no problem.

brightskye committed:
fix disabled gpu information message format

papampi reviewed (approved):
All looks fine, and no errors on my 3 card test rig, hopefully no problem on +10 cards too.
@LuKePicci, @leenoox You see any problems?

papampi commented:
Thanks a lot @brightskye.

LuKePicci commented on helpers/disabled_gpu:
```diff
@@ -0,0 +1,133 @@
+    echo "DISABLED_GPUS not set"
+else
+    # Export CUDA_DEVICE_ORDER env variables ??? not sure if this work
+    export CUDA_DEVICE_ORDER=PCI_BUS_ID
```
As long as this helper script is included into 0miner by `source` command this variable will be exported to any child process of 0miner, an therefore any miner it launches.

papampi commented:
@brightskye , would you please add dev revision number and info to changed files too.

LuKePicci reviewed (changes requested):
Almost fine, just need a little more integration:
- Try to postpone the call to the helper functions and inclusion where and only if this functionality is enabled.
- Instead of injecting the device list at the end of miner specific OPTS variable, which do not exist for every miner, inject it into extension arguments.  You can call the build_disabled_opt(FMT) function to perform the injection just before every line starting with "HCD=miner command line", passing as FMT argument a constant defining the format. Define those constants inside the helper script outside any function.
- Do something similar for command lines of miners defined for salfter's switchers. You can simply append $(build_disabled_opt(FMT)) in the end of any command line inside all json trees.
- Add dev info and version numer taking inspiration from other files like 3main or 0miner, add support for reporting this version numbers in nvOC report command, better to initiate a new "helpers" section for such kind of shared code units.
- Clean up commented code and give pass it undes shellecheck.

leenoox commented:
If you already coded this (disable certain GPU's) you might as well add the
long requested feature of multi miners since lots of the code I believe can
be reused for multi. Besides "disable" add option "multi", eg. GPU's
0,1,2,3 to mine ETH, GPU's 4,5,6 to mine ZEC. Or is it not related and
we'll just open a new can of worms?

On Tue, Jul 24, 2018, 11:06 AM Luca Piccirillo, <notifications@github.com>
wrote:

> *@LuKePicci* requested changes on this pull request.
>
> Almost fine, just need a little more integration:
>
>    - Try to postpone the call to the helper functions and inclusion where
>    and only if this functionality is enabled.
>    - Instead of injecting the device list at the end of miner specific
>    OPTS variable, which do not exist for every miner, inject it into extension
>    arguments. You can call the build_disabled_opt(FMT) function to perform the
>    injection just before every line starting with "HCD=miner command line",
>    passing as FMT argument a constant defining the format. Define those
>    constants inside the helper script outside any function.
>    - Do something similar for command lines of miners defined for
>    salfter's switchers. You can simply append $(build_disabled_opt(FMT)) in
>    the end of any command line inside all json trees.
>    - Add dev info and version numer taking inspiration from other files
>    like 3main or 0miner, add support for reporting this version numbers in
>    nvOC report command, better to initiate a new "helpers" section for such
>    kind of shared code units.
>    - Clean up commented code and give pass it undes shellecheck.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/papampi/nvOC_by_fullzero_Community_Release/pull/155#pullrequestreview-139937367>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Ags0ByJaif-aYKUMxEBD4CkhBQRyVWNuks5uJzf9gaJpZM4VaH3G>
> .
>

papampi commented:
I even think we should merge this PR and later additions to 19-3.1, finalize the 19-2.1 beta so we can release 19-3.0 stable .

papampi commented:
@leenoox , I totally forgot about multi miners, good to brought it up.

papampi commented:
With the pace we are adding more stuff, we will never get to release 19-3 stable

brightskye committed:
Merge branch '19-2.1' into disabled-gpu

brightskye committed:
clean up comments, added Z_EWBF support and added change log comments

brightskye committed:
Clean up more comments

LuKePicci commented:
@papampi yes, this one is a good candidate as first feature for the next 3.1 dev branch

brightskye commented:
@LuKePicci I'm not too sure about using a different extension arguments. I think of these as helpers, which i tried to make it less intrusive.  Experienced users can disabled the gpu by setting each miner's device opts, this script is merely a more convenient tools that helps the user to update their 1bash's device opts setting
As for salfer, shouldn't it be updated to include these additional miner-opts variable since it seems to be the new direction which doesn't exist prior 19-2.1
As for nvoc reporting, i'm a bit lost, just let me know what i need to do.

LuKePicci commented:
I completely agree on your point, I don't want you to define new extension arguments, just inject at runtime as you did into existing coin extension arguments instead of BMINER_OPTS, ETHMINER_OPTS, CLAYMORE_OPTS, etc. Those "miner extension argument" are not defined generally, they only exist for a bunch of them. So, without touching 1bash at all, just append the generated device opt with list in ${!xargs} instead of  CLAYMORE_OPTS and equivalent variables. Would be totally transparent and if you disable the functionality won't break anything. Same for salfter's, but ${!xargs} is not available in there, so you have to append some $DEVICES to command lines.

Let me know which constants you intend to use for device opt format.

About the reporting, it si pretty simple, there is a "nvOC report" command that generate reports with file version and useful info, just add your helper version as it has been done for other scripts and report the value of DISABLE_GPU value.

papampi committed:
Merge branch '19-3.1' into disabled-gpu

LuKePicci commented:
@brightskye if you would like to avoid defining constants for each device opt format you can also generate it dynamically using regex with something like that: https://www.mail-archive.com/python-list@python.org/msg125198.html

So you would have, for ccmienr, 
`build_device_opt "--devices (D,){\$((N-1))}D" `
 and it will return "--devices 2,5,6" assuming those three gpus are enabled.

Inside build_device_opt you would just replace N with enabled gpu count, evaluate the expression using `eval` such that the input string becomes "-- devices (D,){2}D" and passthis as regex to the script I linked, it will generate "--devices D,D,D", finally for every enabled gpu index replace an occurrence of 'D' with that index.

This will likely make build_device_opt implementation general for any miner supporting even those which do not exist yet. With no surprise, pluggable mienrs can have the argument for build_device_opt c bundled within the miner manifest.

brightskye commented:
I see where you are coming from now. I was looking at your plugable miner implementation.
I'll give it a thought as of what can be done since this has been moved to 19-3.1 which gives me more time to think.

LuKePicci commented:
Sure, take all the time you need.

brightskye committed:
refactor disable_gpu.sh
rename disable_gpu to disable_gpu.sh
added generic_helper.sh for general helper

brightskye committed:
merge from 19-3.1

brightskye committed:
add miner to dgh_get_miner_opts $1 for known miner

brightskye committed:
Add hard coded miner opts MDPA MDTA MDDA definition
Fix && || logic errors
change default delimiter for devices to empty string

brightskye committed:
update disabled_gpu script to .sh extension

brightskye committed:
update disabled_gpu script to .sh extension
update miner_opts for each miner

brightskye committed:
merge with 19-3.1

brightskye committed:
renamed disabled_gpu to disabled_gpu.sh

brightskye commented:
I have rewrote quite some part of the code, though there're still improvement that can be made, i find it hard to move forward without a clearer picture.
I think most of the part should be working, though I can't test the final integration. 
Suggestions are welcomed

papampi committed:
Merge pull request #170 from papampi/19-2.1

Merge 19-2.1 to 19-3.1

papampi committed:
Revert "Merge 19-2.1 to 19-3.1"

papampi committed:
Merge pull request #171 from papampi/revert-170-19-2.1

Revert "Merge 19-2.1 to 19-3.1"

LuKePicci commented:
I will give it a careful look asap, thank you!

LuKePicci commented on 0miner:
```diff
@@ -149,68 +154,76 @@ fi
 if [[ $ALGO == ETHASH ]]
 then
 ## Declare Miner Start
+  miner_opts="$(dgh_enabled_devices "${MDPA[$ETHASH_MINER]}" "${MDDA[$ETHASH_MINER]}" "${MDTA[$ETHASH_MINER]}" "$(dgh_get_miner_opts "$ETHASH_MINER" "$COIN" "$ALGO")")"
```
if you change this line as follows you can put is once before the "if ALGO ==" tree:
```
miner_opts="$(dgh_enabled_devices "${MDPA[${!xminer}]}" "${MDDA[${!xminer}]}" "${MDTA[${!xminer}]}" "$(dgh_get_miner_opts "${!xminer}" "$COIN" "$ALGO")")" 
```

brightskye replied:
Actually dgh_get_miner_opts is capable of dyanmic look up which GLOBAL variable to be used by passing it $COIN and $ALGO, i'm restricting the use of it in 0miner because of how 0miner is building the command. Don't wish to introduce more head spinning.
IF you look at the description of dgh_get_miner_opts, without passing in the miner value, it will check GLOBAL VARIABLE in the ORDER of ${COIN}_MINER, ${ALGO}_MINER to retrieve the defined $miner, then check in the order of ${COIN}_${miner}_OPTS, ${ALGO}_${miner}_OPTS, ${miner}_OPTS to try to figure out which OPTS to be used, it's built this way having in mine that you can assign different GPU for running different ALGO/COIN, which I think will prepare you guys for multi miner implementation

brightskye replied:
ETHASH_MINER is "hardcoded in the ETHASH if condition is because I'm afraid to introduce too much flexibity and cause hard to test bugs in 0 miner

LuKePicci replied:
I've looked some parts of your implementation, that's why I say you can simply assign miner_opts outside that if block. You have to take into account that the entire "if ALGO ==" block is likely to be dropped as soon as builtin miners get deprecated. If you need the miner name that is into ${!xminer}, will be ETHMINER or whatever when ALGO==ETHASH and COIN=ETH. Obviously the implementation should be transparent: meanwhile you have no support for general miners, you have to handle unsupported miners as well by falling back to some predefined behavior, like all gpus enabled for example.
To support a general mniner you would take as input something that describes the miner option argument format, not the miner name, but for now it is ok not to have such a mechanism, so take the miner name from ${!xminer}

brightskye replied:
I have moved  miner_opts out from ETHASH, ZHASH and EQUIHASH condition. However, the algo using CCMINER remained, due to how it's defined. IMHO we should leave it like that for the time being I'm reluctant to spend too much time making assumption. I will revisit it when it's clearer on what changes will be made to the miners related  implementation down the road. What do you think?

Just to clarify,
My intention is not to have the disabled-gpu script aware of each miner's information, but a  function to parse and manipulate string based on arguments passed in and the related global variable defined in 1bash (in this case miner related_OPTS). The script is made not to find the dev prefix and GPU # format like how it can find the opts, because I think those can get defined in the json file once your portable miner is implemented. The prefix and gpu format is very miner dependent and it can be easily defined when you defined a new miner.
The global variable MDPA, MDDA, MDTA should be removed after built in miners get deprecated, they are just there for the transition period.

brightskye replied:
"you have to handle unsupported miners as well by falling back to some predefined behavior, like all gpus enabled for example."

There's really no fall back for unknown miners, this totally depends on what the miner accept for it to work. if we don't know a miner, making assumption of the prefix might even break the miner, might as well do nothing :)

brightskye committed:
moved miner_opts out from some of if condition

LuKePicci commented:
The fallback assumption is that if you doadd the device select option at all the miner should enable all gpus. Do you think it is not reasonable?

Don’t worry, take all the time you need to think how to generalize this completely.

________________________________
Da: Victork <notifications@github.com>
Inviato: Tuesday, July 31, 2018 11:14:28 PM
A: papampi/nvOC_by_fullzero_Community_Release
Cc: Luca Piccirillo; Mention
Oggetto: Re: [papampi/nvOC_by_fullzero_Community_Release] Disabled gpu (#155)


@brightskye commented on this pull request.

________________________________

In 0miner<https://github.com/papampi/nvOC_by_fullzero_Community_Release/pull/155#discussion_r206685979>:

> @@ -149,68 +154,76 @@ fi
 if [[ $ALGO == ETHASH ]]
 then
 ## Declare Miner Start
+  miner_opts="$(dgh_enabled_devices "${MDPA[$ETHASH_MINER]}" "${MDDA[$ETHASH_MINER]}" "${MDTA[$ETHASH_MINER]}" "$(dgh_get_miner_opts "$ETHASH_MINER" "$COIN" "$ALGO")")"


"you have to handle unsupported miners as well by falling back to some predefined behavior, like all gpus enabled for example."

There's really no fall back for unknown miners, this totally depends on what the miner accept for it to work. if we don't know a miner, making assumption of the prefix might even break the miner, might as well do nothing :)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/papampi/nvOC_by_fullzero_Community_Release/pull/155#discussion_r206685979>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AIUXtn_4frsfpSDm7uKDpbgq2bvvgoflks5uMMi0gaJpZM4VaH3G>.

brightskye commented:
"The fallback assumption is that if you don't add the device select option at all the miner should enable all gpus. Do you think it is not reasonable?"
I believe it's already doing this?

I'm a bit confused now.
Let's unify our language for a bit.
dgh_enabled_devices - 1 of the main function of the script, takes 4 arguments,
 - $1 (prefix) - prefix to specify when enabling device optison, e.g. --cuda_devices 
 - $2 (delimiter) - seperator between gpus presented after prefix, default to empty string
 - $3 (format) - GPU#  should be presented as numeric or alphanumeric, default to numeric
 - $4 (miner_opt) - additional options for miners, defined in 1bash
behaviour of this dgh_enabled_devices
 1) no miner_opt is presented, it will return a string with all gpus listed minus the blacklisted and format 
     according to prefix, delimiter, format. 
     a) E.g. prefix='--devices ', gpus=5, no disabled gpu, u get #--devices 01234 
     b) E.g. prefix='', #01234
     c) E.g. prefix='', DISABLED_GPU='3' #0124
2) miner_opt is presented, no matching prefix,
    a) prefix='', gpus=5, miner_opts='--api 127.0.0.1' # --api 127.0.0.1 01234
    b) prefix='--device' #--api 127.0.0.1 --device 01234
    c) E.g. prefix='--device', DISABLED_GPU='3' #--api 127.0.0.1 --device 0124
3) miner_opt is presented, matching prefix
    a) prefix='--device', gpus=5, miner_opts='--api 127.0.0.1 --device 134' # --api 127.0.0.1 --device 134
    b) prefix='--device' DISABLED_GPU='3'#--api 127.0.0.1 --device 14

I think this pretty much cover what we need, with the exception of 2.a, whether we should return gpu# without prefix defined

brightskye commented:
So, next step is how _dgh_enabled_devices_ should be called, presumably from 0miner.
Currently, I don't have a good way to call this function from 0miner not all miner is launch in a unified a pattern. So I added the MDPA, MDDA and MDTA for some known miners, however since these are associative arrays, and calling MDPA[$unknow variable] might cause it to just returned the first defined value in the array and result in unexpected behavior, which is why I hesitated.

brightskye commented:
So... is _dgh_enabled_devices_ flexible enough to cover the fall back u mentioned? 
If you are interested, I also have some test cases for the 2 scripts i wrote at 
https://github.com/brightskye/nvoc-fork/tree/disabled-gpu

So.. my thought is to wait for the portable/plugin miners implemented, and refactor 0miner then evaluate how to properly call _dgh_enabled_devices_ to allow maximum coverage?

LuKePicci commented:
You have defined to different functions:
- `dgh_enabled_devices` is general enough but it needs as input something that is miner-specifric, and it is ok, so ww need to push its call down to the inner if branches 
 - `dgh_get_miner_opts` is general and it does not depend o miner-specific inputs, so we can call it immediately

So, from the two previous considerations, you can do the following.

Before any if block
`miner_opts=$(dgh_get_miner_opts "${!xminer}" "$COIN" "$ALGO")")`
its implementation is ok, it handles each builtin or pluggable miner names just fine

Then, since `dgh_enabled_devices` needs miner-specifc arguments has to be called at some point in the code where we have such information available. Your associative array approach takes as input the miner name to bind relevant prefix, delimeter, format. You don't need such an array, for example you can simply do something like that:

```
## ETHASH
if [[ $ALGO == ETHASH ]]
then
## Declare Miner Start
  if [[ $ETHASH_MINER == GENOIL ]]
  then
    miner_opts="$miner_opts $(dgh_enabled_devices "" "" "" "$miner_opts")"
    HCD="${NVOC}/miners/Genoil-U/ethminer -S ${!xpool}:${!xport} -O ${!xaddr}${!xwallet}${!xwork}:$MINER_PWD -SP 1 -U ${!xext} $miner_opts"
  elif [[ $ETHASH_MINER == ETHMINER ]]
  then
    miner_opts="$miner_opts $(dgh_enabled_devices "--cuda_devices" " " "n" "$miner_opts")"
    HCD="${NVOC}/miners/ethminer/latest/ethminer -S ${!xpool}:${!xport} -O ${!xaddr}${!xwallet}${!xwork}:$MINER_PWD -SP 1 -U ${!xext} $miner_opts"
  elif [[ $ETHASH_MINER == CLAYMORE ]]
  then
...
```
This is the best way of keeping miner-specific details inside the miner-specific if branch, and it is also perfectly compatibile with the pluggable miner implementation (it has been merged already, just two more 'if' branches have been added at the beginning, there should be no more refactor if nothing to be changed comes up, take a look and ask me if something in the code is still not clear) which takes care of reading the thee argument for `dgh_enabled_devices` from the pluggable miner's manifest and use it to in the same way as I showed you above for ethminer and genoil. Passing empty prefix (like I did here for genoil), if I understood correctly, should allow all gpus to be enabled (this resembles the "unknown/unsupported miner") so this is pretty fine.

This should be enough to cover all cases. You can safely drop the associative array, it is not needed.

LuKePicci commented:
In my previous example I assumed `dgh_enabled_devices` output contains just the opt argument related to device selection and not the whole options set.

brightskye committed:
removed MINER OPTS CONSTANT MDDA, MDTA, MDPA
use direct input of each prefix, delimiter and type at 0miner

brightskye committed:
Merge branch '19-3.1' into disabled-gpu

brightskye commented:
Sorry for the delayed,
@LuKePicci, I just made the changes as you suggested, let me know if it's okay this time

LuKePicci commented:
I know that if you skip calling that function the behavior is already as I described, but in pluggable miners I have to call it with unforeseeable values taken from the miner manifest, for example, assume I check that those arguments for `dgh_enabled_devices` are available in the miner manifest, then we call that function parsing those arguments, we have to make sure that `dgh_enabled_devices` silently return no device select option in case any argument is bad. Consider tha case where someone used wrong values in the manifest such that the function call looks like:

```
miner_opts="$(dgh_enabled_devices 'blabla' ':-P' '?' "$miner_opts")" 
```

That should possibly result in no errors or extra (bad) arguments inserted. Is this already the case?

brightskye commented:
ha, this is a tough question, from what i remember
"$(dgh_enabled_devices 'blabla' ':-P' '?' "$miner_opts")"  will return something like 
# blabla0:-P1:-P2:-P3 
# assuming miner_opts was empty previously
#or blabla0:-P1:-P2:-P3 will be appended at the end of the miner_opts, i think it's save as long as the #miner don't  recognize it

brightskye committed:
removed double quote to avoid reading miner_opts as single argument

brightskye commented:
I'm not sure about this, i thought eval will combine all argument as string, and evaluate them, so it doesn't matter if it's double quoted. anyway, i will remove the double quote

brightskye committed:
added miner_opts to remaining command line

LuKePicci commented:
Ok, I will try to sort out the issue with random arguments somewhere else.

brightskye commented:
I'm not sure if you need to worry about random argument too much
IMHO GIGO is perfectly fine, i think it's better to be transparent to the user, instead of trying to do too much abstraction.
As long as it's easy to debug, e.g. if they put in some weird value, it will be printed on the command line and they can easily examine the command line to figure what's wrong

LuKePicci commented:
Yeah, you're right, we just need something to allow users not to specify this set of arguments withouut breaking things. Avoiding to call `dgh_enabled_devices` in case the first one is empty should be enough.

papampi commented:
Sorry was working on Dual_Cuda new image, completely forgot about this.
Is this finished?

LuKePicci commented:
Yes, this is finished but some conflicts needs to be resolved against the 3.1 branch.

LuKePicci commented:
@brightskye if you have a bit of time to take care of resolving conflicts do it as it is a good moment to merge this in 3.1-dev

brightskye committed:
update to upstream 3.1-dev branch

brightskye commented:
Sorry guys, it's been awhile and I haven't been paying attention here.
Please check and merge it back to 3.1-dev
Thanks,

LuKePicci reviewed (changes requested):
I see mamy code blocks being re-added from old files versions as they were at the time of coding your feature, make sure this pr does only include changes relative to this feature.

brightskye committed:
fix previous merge issue

brightskye commented:
done, please check

LuKePicci commented:
Looks good now. @papampi give a look and merge, for me it is fine, any eventual issues will be solved when we manage to test this properly.

papampi commented:
looks good, should be tested after merge

papampi commented:
@brightskye 
Can you please check this : https://github.com/papampi/nvOC_miners/issues/34
